### PR TITLE
feat:add hdfs support

### LINKED ISSUE #236: Add components for file interface
**What would you like to be added:**
 
Layotto support file interface now, and have implemented the [alioss](https://help.aliyun.com/product/31815.html?spm=5176.22414175.sslink.2.19d9d8a6z4eRDE) component.

So this issue want to support other file system, eg, 

- [x] [awsOss](https://aws.amazon.com/cn/campaigns/free/?sc_channel=PS&sc_campaign=acquisition_CN&sc_publisher=google&sc_category=pc&sc_medium=Google_%E5%93%81%E7%89%8C_%E6%A0%B8%E5%BF%83%E5%93%81%E7%89%8C%E8%AF%8D_b&sc_content=%E6%A0%B8%E5%BF%83%E5%93%81%E7%89%8C%E8%AF%8D&sc_detail=aws&sc_segment=2021092300256&sc_matchtype=exact&sc_country=CN&all-free-tier.sort-by=item.additionalFields.SortRank&all-free-tier.sort-order=asc&awsf.Free%20Tier%20Types=*all&awsf.Free%20Tier%20Categories=*all) assigned to @zach030
- [ ] local file system(file on local disk)
- [ ] [HDFS](https://hadoop.apache.org/)
- [x] [minIo](http://www.minio.org.cn/)  assigned to @zach030
- [ ] ftp
- many other file systems...

You can refer the interface design document of file [component interface design](https://mosn.io/layotto/#/zh/design/file/file-design),[api design](https://mosn.io/layotto/#/zh/api_reference/file/file)

### PR DESCRIPTION
**Pull request by bokket**
Signed-off-by: bokket <3100563328@qq.com>

**What this PR does**:
add Hdfs oss support for Layotto file system interface.

**Which issue(s) this PR fixes**:

Fixes #236

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by seeflood** [comment]
@wenxuwan Hi,how is the review process going on ? Can we merge this PR?

**Comment by wenxuwan** [comment]
@bokket hi man, can you fix the conflict? and adapt you implementation for the new file interface?

**Comment by bokket** [comment]
> @bokket hi man, can you fix the conflict? and adapt you implementation for the new file interface?

Hi.I will fix it in next week.

**Comment by wenxuwan** [comment]
> > @bokket hi man, can you fix the conflict? and adapt you implementation for the new file interface?
> 
> Hi.I will fix it in next week.

hi man, any update?

**Comment by bokket** [comment]
> > > @bokket hi man, can you fix the conflict? and adapt you implementation for the new file interface?
> > 
> > 
> > Hi.I will fix it in next week.
> 
> hi man, any update?

Sorry for the delay in my response.Recent course design time is tight.My local pseudo-distribution passed the test, but it seems that the reason for failure in CI is related to `go.sum` change, can you give me some hints?

**Comment by seeflood** [comment]
hi @bokket, you can type `go mod tidy` to fix the CI problem.
Besides, remember to fix the git conflicts :)

**Comment by seeflood** [comment]
It seems the conflicts have been solved :)
@wenxuwan Could u please help review it again?

File: components/file/hdfs/Hdfs.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+	it, err := client.List(starter)
+	if err != nil {
+		err = ErrHdfsListFail
+	}
+
+	resp.Marker = it.ContinuationToken()
+	for {
+		o, err := it.Next()
+		if err != nil && !errors.Is(err, types.IterateDone) {
+			fmt.Errorf("fail to List All Next: %v", err)
```
Comment (wenxuwan) [inline components/file/hdfs/Hdfs.go:160]: err occur, should return error directly?

**Comment by wenxuwan** [comment]
@bokket hi man, any progress？

**Comment by bokket** [comment]
> @bokket hi man, any progress？

If you are Satisfied with this pr it should be ready to merge.I will try to set up a github CI for this change over winter break.

**Comment by wenxuwan** [comment]
> > @bokket hi man, any progress？
> 
> If you are satisfied with this pr it should be ready to merge.I will try to set up a github CI for this change over winter break.

List method still have some problem, you can check my comments ,and refer https://github.com/mosn/layotto/blob/8be35c7b054a16137db0033abb4fa31ff8acf787/components/file/s3/aws/oss.go#L182

**Comment by seeflood** [comment]
I modified the pipeline to test components and sdk packages in #367 .
Please check the hdfs ut failure  :)
@bokket

File: components/file/hdfs/Hdfs.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+	starter := ""
+	resp := &file.ListResp{}
+
+	it, err := client.List(starter)
+	if err != nil {
+		err = ErrHdfsListFail
+	}
+
+	resp.Marker = it.ContinuationToken()
+	for {
```
[the remark this answers is not in the document]
Comment (bokket) [reply components/file/hdfs/Hdfs.go:156]: Marker is represented a key assigned to an object to retrieve.`o.path`is the base name of the file's.I think they are the same when working with large files.

**Comment by bokket** [comment]
> I modified the pipeline to test components and sdk packages in #367 . Please check the hdfs ut failure :) @bokket

These changes do not seem to be any problem, but it is strange that CI failed when testing to hdfs....

**Comment by seeflood** [comment]
@bokket The CI pipeline didn't run tests for components package,so we didn't know the ut failure until I modified the pipeline configuration.
You can run your hdfs ut locally to check the failure. 
Here's the error message I reproduced in my goland IDE:
image

**Comment by bokket** [comment]
> 
image
then it may be `action.yaml` did not add the port to the configuration, I tried to add hdfs port above `ut`

**Comment by seeflood** [comment]
Thanks for your contribution!
I commented out the hdfs related configuration in `github-actions.yml` because it's not in use right now

**Review by wenxuwan — APPROVED** [review]
LGTM

**Review by seeflood — APPROVED** [review]
lgtm

**Pull request merged**